### ISSUE TITLE
bug: discrepancy between buffers in `runtime.go` and `runtime/bufferpool.go`

### ISSUE THREAD

**Issue by JakubCzarlinski** [a-h/templ#903]
Hey All,

I was profiling a project that I'm working on an noticed the buffers used are inconsitent.

Here the `bytes.Buffer` is used and allocates a minimum of `const smallBufferSize = 64`.

https://github.com/a-h/templ/blob/c7c32aa5388f559fe3684b8bae967de46909b37f/runtime.go#L570-L583

Here the `bufio.Buffer` is used and allocates a default `var DefaultBufferSize = 4 * 1024 // 4KB`

https://github.com/a-h/templ/blob/c7c32aa5388f559fe3684b8bae967de46909b37f/runtime/bufferpool.go#L8-L38

This causes extra `grow` operations on the buffers used in `runtime.go`.

**Comment by a-h** [comment]
The `runtime` package is newer, and is now used for buffering responses if you use a recent version of templ to generate code.

The new version allows flushing partial responses to the underlying stream, whereas the previous version held the hold response in RAM before flushing.

I would expect them to behave differently, and have different amounts of buffer sizes, so I don't see this as a bug.

Am I missing the point of the issue?

Is there a problem with the performance of either that you're seeing in production or during load testing?

Thanks!

**Comment by JakubCzarlinski** [comment]
If you do something like `http.Handle("/", templ.Handler(component))` then from what I understand the smaller buffer is still used. Is this intended?

Context: https://templ.guide/quick-start/running-your-first-templ-application

**Comment by a-h** [comment]
Yes, that's right. The complete response is buffered to prevent partial responses from going to the client.

templ templates can fail to render part way through, for example, if `{ funcThatReturnsAStringAndError() }` is used. In that case, the rendering will stop.

Ideally, you'd want a HTTP 500 error at this point, but... HTTP status codes must be sent before the body is sent.

So, to avoid a situation where you get a partial response, but a HTTP 200 status code, by default `templ.Handler` renders the whole response to an temporary internal buffer, and if that's successful, it copies the buffer to the output.

Alternatively, you can use the `WithStreaming` option in the handler to avoid the buffering https://github.com/a-h/templ/blob/3ac3c9de53ad81f43d70bebfb642a89c2c69a4c1/handler.go#L40-L53
